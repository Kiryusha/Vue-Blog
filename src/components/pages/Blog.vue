<template lang="pug">
  .blog
    Delete(
      @fetchList="fetchList",
      @fetchCategories="fetchCategories"
    )
    .container
      .blog__content
        section.blog__inner
          router-view(
            :list="list"
          )
        section.blog__sidebar
          SidebarBlock(v-if="categories.length")
            span(slot="title") Категории
            Categories(
              :categories="categories",
              @fetch="fetchList"
            )
</template>

<script>
import axios from 'axios';
import SidebarBlock from '../elements/blog/SidebarBlock';
import Categories from '../elements/blog/Categories';
import Delete from '../elements/modals/Delete';

export default {
  components: {
    SidebarBlock,
    Categories,
    Delete,
  },
  data() {
    return {
      list: [],
      categories: [],
    };
  },
  mounted() {
    this.$Progress.finish();
    this.fetchList();
    this.fetchCategories();
  },
  methods: {
    fetchList(category) {
      this.$Progress.start();

      let api = '/api/posts/';

      if (category) {
        api = `/api/categories/${category}/`;
      }

      axios.get(api).then((response) => {
        this.list = response.data;
        if (category && this.$route.path !== '/blog/') {
          this.$router.push({ path: '/blog/' }, () => {
            this.$Progress.finish();
          });
        } else {
          this.$Progress.finish();
        }
      });
    },
    fetchCategories() {
      this.$Progress.start();

      axios.get('/api/categories/').then((response) => {
        this.categories = response.data;
        this.$Progress.finish();
      });
    },
  },
};
</script>

<style lang="stylus">
  .blog
    min-height 100%
    padding-top 60px
    padding-bottom 250px
    background-attachment fixed
    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 80 80' width='80' height='80'%3E%3Cpath fill='%239C92AC' fill-opacity='0.4' d='M14 16H9v-2h5V9.87a4 4 0 1 1 2 0V14h5v2h-5v15.95A10 10 0 0 0 23.66 27l-3.46-2 8.2-2.2-2.9 5a12 12 0 0 1-21 0l-2.89-5 8.2 2.2-3.47 2A10 10 0 0 0 14 31.95V16zm40 40h-5v-2h5v-4.13a4 4 0 1 1 2 0V54h5v2h-5v15.95A10 10 0 0 0 63.66 67l-3.47-2 8.2-2.2-2.88 5a12 12 0 0 1-21.02 0l-2.88-5 8.2 2.2-3.47 2A10 10 0 0 0 54 71.95V56zm-39 6a2 2 0 1 1 0-4 2 2 0 0 1 0 4zm40-40a2 2 0 1 1 0-4 2 2 0 0 1 0 4zM15 8a2 2 0 1 0 0-4 2 2 0 0 0 0 4zm40 40a2 2 0 1 0 0-4 2 2 0 0 0 0 4z'%3E%3C/path%3E%3C/svg%3E");

    &__content
      display flex
      justify-content space-between
      align-items flex-start

    &__inner
      flex-grow 1
      max-width 100%

    &__sidebar
      flex-basis 270px
      min-width 270px
      margin-left 40px
      max-width 100%

    +mobile()
      padding-top 40px

      &__content
        flex-direction column-reverse
        align-items inherit

      &__inner
        max-width none

      &__sidebar
        flex-basis 0
        margin 0 0 40px

    +phone()
      padding-top 20px

      &__sidebar
        margin 0 0 20px
</style>
