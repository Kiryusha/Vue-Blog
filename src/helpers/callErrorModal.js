// calling error modal
export default function (context, error) {
  const errText = error.response ?
    error.response.data.message : error.message;
  context.$modal.show('response', { message: errText });
  context.$Progress.fail();
}
