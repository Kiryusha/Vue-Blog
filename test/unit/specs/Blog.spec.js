// Blog.vue

import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import axios from 'axios';
import VueAxios from 'vue-axios';
import Blog from 'Components/pages/Blog.vue';
import installUnhandledPromise from 'jest-plugin-unhandled-promise';
import callErrorModal from '@/helpers/callErrorModal';

// Jest, since v21, should do it yourself, but I didnt find that option.
installUnhandledPromise(console.log);
// Unless we connect axios and vue-axios, Jest cannot mock "this.axios"
Vue.use(VueAxios, axios);

jest.mock('axios', () => ({
  get: jest.fn(),
}));

jest.mock('@/helpers/callErrorModal', () => jest.fn());

describe('Blog.vue', () => {
  // $route.fullPath is used as :key value in router-view
  const initBlog = jest.fn();
  const fetchCategories = jest.fn();
  const fetchList = jest.fn();
  const addScroll = jest.fn();
  const removeScroll = jest.fn();
  const endlessScroll = jest.fn();
  let wrapper;
  let config =  {
    stubs: ['router-view'],
    mocks: {
      $route: {
        fullPath: 'full/path',
        path: '/blog/',
      },
      $router: {
        push: jest.fn(),
      },
      $modal: {
        show: () => {},
      },
      $Progress: {
        fail: () => {},
      },
    },
  };

  beforeEach(() => {
    axios.get.mockClear();
    axios.get.mockResolvedValue({ data: {} });
  });

  afterEach(() => {
    jest.clearAllMocks();
    config.mocks.$route.path = '/blog/';
  });

  describe('component shallow rendering', () => {
    beforeEach(() => {
      wrapper = shallowMount(Blog, config);
    });

    it('matches shallow snapshot of itself', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders sidebar if has categories', () => {
      wrapper.setData({ categories: [1, 2, 3] });

      expect(wrapper.contains('sidebarblock-stub')).toBe(true);
    });
  });

  describe("lifecycle-driven methods calls", () => {
    it('calling initBlog() on a created hook', () => {
      const wrapper = shallowMount(Blog, {...config, methods: {
        initBlog,
      }});

      expect(initBlog).toHaveBeenCalledTimes(1);
    });

    it('calling fetchCategories() on a created hook', () => {
      const wrapper = shallowMount(Blog, {...config, methods: {
        fetchCategories,
      }});

      expect(fetchCategories).toHaveBeenCalledTimes(1);
    });

    it('fetchCategories() should set categories to API value', async () => {
      const data = ['cat', 'dog'];
      axios.get.mockResolvedValue({ data });

      const wrapper = shallowMount(Blog, config);
      await wrapper.vm.$nextTick();

      expect(axios.get).toHaveBeenCalledWith('/api/categories/');
      expect(wrapper.vm.categories).toBe(data);
    });

    it('calling fetchList() on a created hook', () => {
      const wrapper = shallowMount(Blog, {...config, methods: {
        fetchCategories,
        fetchList,
      }});

      expect(fetchList).toHaveBeenCalledTimes(1);
    });

    it('fetchList() should set list to API value', async () => {
      const data = [{ cat: 'dog' }, { cat: 'dog' }, { cat: 'dog' }];
      axios.get.mockResolvedValue({ data });

      const wrapper = shallowMount(Blog, {...config, methods: {
        fetchCategories,
      }});
      await wrapper.vm.$nextTick();

      expect(axios.get).toHaveBeenCalledWith('/api/posts/1/');
      expect(wrapper.vm.list).toBe(data);
    });

    it('calling asynchronously addScroll() on a created hook', async () => {
      const wrapper = shallowMount(Blog, {...config, methods: {
        fetchCategories,
        fetchList,
        addScroll,
      }});
      await wrapper.vm.$nextTick();

      expect(addScroll).toHaveBeenCalledTimes(1);
    });

    it('addScroll() attaches endlessScroll() to the window scroll event', async () => {
      const wrapper = shallowMount(Blog, {...config, methods: {
        fetchCategories,
        fetchList,
        endlessScroll,
      }});
      await wrapper.vm.$nextTick();
      global.window.dispatchEvent(new Event('scroll'));
      wrapper.destroy();

      expect(endlessScroll).toHaveBeenCalled();
    });

    it('calling removeScroll() on a destroyed hook', () => {
      const wrapper = shallowMount(Blog, {...config, methods: {
        addScroll,
        removeScroll,
      }})
      wrapper.destroy();

      expect(removeScroll).toHaveBeenCalledTimes(1);
    });

    it('removeScroll() removes endlessScroll() from the window scroll event', async () => {
      // Attaching events to global objects is hell for unit testing
      const wrapper = shallowMount(Blog, {...config, methods: {
        fetchCategories,
        fetchList,
        endlessScroll,
      }})
      await wrapper.vm.$nextTick();

      wrapper.destroy();
      global.window.dispatchEvent(new Event('scroll'));

      expect(endlessScroll).not.toHaveBeenCalled();
    });
  });

  describe("custom actions", () => {
    it('fetchList() should add API value to list', async () => {
      const list = [{ gato: 'perro' }];
      const data = [{ cat: 'dog' }];

      axios.get.mockResolvedValue({ data });

      const wrapper = shallowMount(Blog, {...config, methods: {
        initBlog,
      }});

      wrapper.setData({ list });
      // imitating scroll to bottom
      wrapper.vm.fetchList(null, 2, true);
      await wrapper.vm.$nextTick();

      expect(axios.get).toHaveBeenCalledWith('/api/posts/2/');
      expect(wrapper.vm.list).toEqual([...list, ...data]);
    });

    it("fetchList('test') should call '/api/categories/category/1/'", async () => {
      const wrapper = shallowMount(Blog, config);
      wrapper.vm.fetchList('test');

      expect(axios.get).toHaveBeenCalledWith('/api/categories/test/1/');
    });

    it("fetchList('test') should set its argument as activeCategory", async () => {
      const wrapper = shallowMount(Blog, config);
      wrapper.vm.fetchList('test');

      expect(wrapper.vm.activeCategory).toBe('test');
    });

    it("fetchList('test') should call redirect if $route is not '/blog/'", async () => {
      config.mocks.$route.path = '/post/';
      const wrapper = shallowMount(Blog, config);

      wrapper.vm.fetchList('test');
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.$router.push).toHaveBeenCalled();
    });

    it("deletePost('test') should delete post from list by code", async () => {
      const list = [{
        code: 'test',
      }, {
        code: 'test2',
      }];
      const result = [{
        code: 'test2',
      }]
      const wrapper = shallowMount(Blog, {...config, methods: {
        fetchList,
      }});

      wrapper.setData({ list });
      wrapper.vm.deletePost('test');
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.list).toEqual(result);
    });
  });

  describe("error handling", () => {
    it("fetchCategories() reject calls callErrorModal()", async () => {
      axios.get.mockResolvedValue();
      axios.get.mockRejectedValue(new Error('test'));

      const wrapper = shallowMount(Blog, config);
      await wrapper.vm.$nextTick();

      expect(callErrorModal).toHaveBeenCalled();
    });

    it("fetchList() reject calls callErrorModal()", async () => {
      axios.get.mockResolvedValue();
      axios.get.mockRejectedValue(new Error('test'));

      const wrapper = shallowMount(Blog, {...config, methods: {
        fetchCategories,
      }});
      await wrapper.vm.$nextTick();

      expect(callErrorModal).toHaveBeenCalled();
    });
  });
});
