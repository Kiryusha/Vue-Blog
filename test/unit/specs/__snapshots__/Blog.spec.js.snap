// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog.vue component shallow rendering matches shallow snapshot of itself 1`] = `
<div class="blog">
  <delete-stub></delete-stub>
  <div class="container">
    <div class="blog__content">
      <section class="blog__inner">
        <router-view-stub list="[object Object]"></router-view-stub>
      </section>
      <section class="blog__sidebar">
        <!---->
      </section>
    </div>
  </div>
</div>
`;
